packages/protocol/contracts/L1/libs/LibVerifying.sol

```diff
@@ -151,24 +172,76 @@ library LibVerifying {
         }
     }
 
-    function _markBlockVerified(
+    function _verifyBlock(
         TaikoData.State storage state,
+        TaikoData.Config memory config,
+        AddressResolver resolver,
         TaikoData.Block storage blk,
-        TaikoData.ForkChoice storage fc,
-        uint24 fcId,
-        address systemProver
+        TaikoData.ForkChoice memory fc,
+        uint24 fcId
     )
         private
     {
-        uint64 proofTime;
-        unchecked {
-            proofTime = uint64(fc.provenAt - blk.proposedAt);
+        // the actually mined L2 block's gasLimit is blk.gasLimit +
+        // LibL2Consts.ANCHOR_GAS_COST, so fc.gasUsed may greater than
+        // blk.gasLimit here.
+        uint32 _gasLimit = blk.gasLimit + LibL2Consts.ANCHOR_GAS_COST;
+        assert(fc.gasUsed <= _gasLimit);
+
+        IProverPool proverPool =
+            IProverPool(resolver.resolve("prover_pool", false));
+
+        if (blk.assignedProver == address(0)) {
+            --state.numOpenBlocks;
+        } else if (!blk.proverReleased) {
+            proverPool.releaseProver(blk.assignedProver);
         }
 
-        uint64 reward;
+        // Reward the prover (including the oracle prover)
+        uint64 proofReward =
+            (config.blockFeeBaseGas + fc.gasUsed) * blk.rewardPerGas;
+
+        if (fc.prover == address(1)) {
+            // system prover is rewarded with `proofReward`.
+        } else if (blk.assignedProver == address(0)) {
+            // open prover is rewarded with more tokens
+            proofReward = proofReward * config.rewardOpenMultipler / 100;
+        } else if (
+            fc.prover == blk.assignedProver
+                && fc.provenAt <= blk.proposedAt + blk.proofWindow
+        ) {
+            // The selected prover managed to prove the block in time
+            state.avgProofDelay = uint16(
+                LibUtils.movingAverage({
+                    maValue: state.avgProofDelay,
+                    // TODO:  prover is not incentivized to submit proof
+                    // ASAP
+                    newValue: fc.provenAt - blk.proposedAt,
+                    maf: 7200
+                })
+            );
+
+            state.feePerGas = uint32(
+                LibUtils.movingAverage({
+                    maValue: state.feePerGas,
+                    newValue: blk.rewardPerGas,
+                    maf: 7200
+                })
+            );
+        } else {
+            // proving out side of the proof window
```

Comment:
If the prover never returns the proof, how do we know it needs to open?

Reply:
If the assigned prover didn't submit a proof within the proving window, then anyone else can submit a proof.